src/goto-instrument/contracts/dynamic-frames/dfcc_cfg_info.h

```diff
@@ -0,0 +1,293 @@
+/*******************************************************************\
+
+Module: Dynamic frame condition checking for function and loop contracts.
+
+Author: Qinheping Hu, qinhh@amazon.com
+Author: Remi Delmas delmasrd@amazon.com
+
+Date: March 2023
+
+\*******************************************************************/
+
+/// \file
+/// Class that computes CFG information about the loop structure
+/// of a GOTO function for the purpose of dynamic frame conditions checking
+/// and loop contract instrumentation.
+
+#ifndef CPROVER_GOTO_INSTRUMENT_CONTRACTS_DYNAMIC_FRAMES_DFCC_CFG_INFO_H
+#define CPROVER_GOTO_INSTRUMENT_CONTRACTS_DYNAMIC_FRAMES_DFCC_CFG_INFO_H
+
+#include <util/message.h>
+#include <util/namespace.h>
+#include <util/std_expr.h>
+
+#include <goto-programs/goto_program.h>
+
+#include "dfcc_loop_contract_mode.h"
+
+#include <map>
+#include <set>
+
+class dfcc_utilst;
+class dfcc_libraryt;
+class goto_modelt;
+class goto_functiont;
+
+/// \brief Describes a single loop for the purpose of DFCC loop contract
+/// instrumentation.
+class dfcc_loop_infot
+{
+public:
+  dfcc_loop_infot(
+    const std::size_t loop_id,
+    const std::size_t cbmc_loop_id,
+    const std::set<exprt> assigns,
+    const exprt invariant,
+    const exprt decreases,
+    const std::set<irep_idt> local,
+    const std::set<irep_idt> tracked,
+    const std::vector<std::size_t> inner_loops,
+    const std::vector<std::size_t> outer_loops,
+    const symbol_exprt write_set_var,
+    const symbol_exprt addr_of_write_set_var)
```

Comment:
All of these need to be reviewed: should they be const references instead? Or actually be movable (in which case you don't want the `const`)?

Reply:
Fixed.